댱뇨병 여부(Outcome)를 예측하는 모델의 성능을 평가하는 것이 전체 분석의 목적입니다.
선택한 이유는 데이터분석 경험을 쌓고자 데이터셋이 크지않고 데이터의 내용 또한 어렵지 않다고 판단되어 이 주제를 선택하였습니다.


In [1]:
import pandas as pd

In [2]:
from pandas import read_csv

In [ ]:
#pandas와 그 중 read_csv 함수를 불러온다. 데이터분석에 용이함

In [3]:
pwd

'/Users/joseong-gwan/Downloads'

In [4]:
import warnings
warnings.filterwarnings('always')
warnings.filterwarnings('ignore')

In [ ]:
# 특정 경고 메시지를 무시하도록 설정 -> 데이터 분석에 집중할 수 있음

In [ ]:
diabetes=pd.read_csv('../input/diabetes/diabetes.csv',na_values=['?'])

In [ ]:
diabetes

In [ ]:
# 데이터를 읽어와서 diabetes라는 변수로 저장 ,na_values=['?']는 데이터셋에서 '?'로 표시된 결측값을 NaN으로 처리하도록 지정

In [6]:
diabetes.describe()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
count,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000
mean,3.845052,120.894531,69.105469,20.536458,79.799479,31.992578,0.471876,33.240885,0.348958
std,3.369578,31.972618,19.355807,15.952218,115.244002,7.884160,0.331329,11.760232,0.476951
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.078000,21.000000,0.000000
25%,1.000000,99.000000,62.000000,0.000000,0.000000,27.300000,0.243750,24.000000,0.000000
50%,3.000000,117.000000,72.000000,23.000000,30.500000,32.000000,0.372500,29.000000,0.000000
75%,6.000000,140.250000,80.000000,32.000000,127.250000,36.600000,0.626250,41.000000,1.000000
max,17.000000,199.000000,122.000000,99.000000,846.000000,67.100000,2.420000,81.000000,1.000000


In [ ]:
# 데이터셋의 통계적 요약을 보여준다.

In [7]:
diabetes.shape

(768, 9)

In [ ]:
# 데이터는 총 768개, 변수는 총 9개

In [ ]:
diabetes.tail() # diabetes의 마지막 5행을 보여준다. 데이터셋의 구조와 내용을 이해할 수 있다.

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
763,10,101,76,48,180,32.9,0.171,63,0
764,2,122,70,27,0,36.8,0.340,27,0
765,5,121,72,23,112,26.2,0.245,30,0
766,1,126,60,0,0,30.1,0.349,47,1
767,1,93,70,31,0,30.4,0.315,23,0


1. 임신기간
2. 포도당 수치
3. 혈압
4. 피부 두께
5. 인슐린 수치
6. 비만 지수
7. 당뇨 유전 확률
8. 나이
9. 당뇨 여부

In [9]:
# Create a pipeline that standardizes the data then creates a model
from pandas import read_csv
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
X = diabetes.loc[:,diabetes.columns!="Outcome"]
y = diabetes.loc[:,'Outcome']
# create pipeline
estimators = []
estimators.append(('standardize', StandardScaler()))
estimators.append(('lda', LinearDiscriminantAnalysis()))
model = Pipeline(estimators)
# evaluate pipeline
seed = 10
kfold = KFold(n_splits=10, random_state=seed)
results = cross_val_score(model, X, y, cv=kfold)
print(results.mean())

0.773462064251538


KFold은 데이터를 여러 번 나눠서 평가하는 것
cross_val_score은 모델 성능을 반복 측정
한 번만 평가하면 운에 따라 결과가 달라질 수 있어서 여러 번 나눠서 신뢰도 높은 성능을 평가합니다.
StandardScaler은 변수들의 단위를 맞추는 것(표준화)
혈압,나이,BMI 등 스케일이 각각 달라서 이대로 사용한다면 특정 변수의 영향이 커지게 됩니다. 변수들의 단위를 맞춤으로써 공정하게 모델을 학습시킵니다.
LDA 모델을 사용해서 분류 분석을 수행합니다. 입력 변수들(X)를 이용해서 Outcome(당뇨 여부)를 예측합니다.
X: 설명 변수(나이, 혈당 등)
Y: 목표 변수(당뇨 여부)
X를 통해 Y를 예측
파이프라인(전처리 + 모델)을 구성
데이터 누수를 방지한다.
랜덤 시드를 설정 seed = 10
결과를 재현 가능하게 만드는 것
데이터를 10개로 나눠서 반복적으로 학슴과 검증을 수행합니다.
모델이 특정 데이터에만 잘 맞는지 확인합니다.
모델 성능을 10번 반복 측정 -> 10개의 정확도 값 (모델이 얼마나 안정적인지 확인)
10개의 정확도 값을 평균을 내보니 약 0.77
새로운 데이터에서도 약 77% 정확도를 기대할 수 있습니다.
0.7~0.8은 보통이거나 괜찮은 성능을 가진다고 말할 수 있습니다.

이 데이터분석의 의의는 당뇨병 예측 모델이 실제 데이터에서도 안정적으로 작동하는지 검증한 분석이라고 볼 수 있습니다.